pkg/chunk/purger/purger.go

```diff
@@ -0,0 +1,576 @@
+		}
+
+		plan := purgeplan.Plan{
+			PlanInterval: &purgeplan.Interval{
+				StartTimestampMs: int64(planRange.Start),
+				EndTimestampMs:   int64(planRange.End),
+			},
+			ChunksGroup: chunksGroups,
+		}
+
+		pb, err := proto.Marshal(&plan)
+		if err != nil {
+			return err
+		}
+
+		plans[i] = pb
+	}
+
+	err := dp.putDeletePlans(ctx, req.UserID, req.RequestID, plans)
+	if err != nil {
+		return err
+	}
+
+	err = dp.deleteStore.UpdateStatus(ctx, req.UserID, req.RequestID, chunk.Deleting)
+	if err != nil {
+		return err
+	}
+
+	level.Info(req.logger).Log("msg", "built delete plans", "num_plans", len(perDayTimeRange))
+
+	return nil
+}
+
+func (dp *DataPurger) putDeletePlans(ctx context.Context, userID, requestID string, plans [][]byte) error {
+	for i, plan := range plans {
+		objectKey := buildObjectKeyForPlan(userID, requestID, i)
+
+		err := dp.storageClient.PutObject(ctx, objectKey, bytes.NewReader(plan))
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (dp *DataPurger) getDeletePlan(ctx context.Context, userID, requestID string, planNo int) (*purgeplan.Plan, error) {
+	objectKey := buildObjectKeyForPlan(userID, requestID, planNo)
+
+	readCloser, err := dp.storageClient.GetObject(ctx, objectKey)
+	if err != nil {
+		return nil, err
+	}
+
+	defer readCloser.Close()
+
+	buf, err := ioutil.ReadAll(readCloser)
+	if err != nil {
+		return nil, err
+	}
+
+	var plan purgeplan.Plan
+	err = proto.Unmarshal(buf, &plan)
+	if err != nil {
+		return nil, err
+	}
+
+	return &plan, nil
+}
+
+func (dp *DataPurger) removeDeletePlan(ctx context.Context, userID, requestID string, planNo int) error {
+	objectKey := buildObjectKeyForPlan(userID, requestID, planNo)
+	return dp.storageClient.DeleteObject(ctx, objectKey)
+}
+
+// returns interval per plan
+func splitByDay(start, end model.Time) []model.Interval {
+	numOfDays := numOfPlans(start, end)
+
+	perDayTimeRange := make([]model.Interval, numOfDays)
+	startOfNextDay := model.Time(((int64(start) / millisecondPerDay) + 1) * millisecondPerDay)
+	perDayTimeRange[0] = model.Interval{Start: start, End: startOfNextDay - 1}
+
+	for i := 1; i < numOfDays; i++ {
+		interval := model.Interval{Start: startOfNextDay}
+		startOfNextDay += model.Time(millisecondPerDay)
+		interval.End = startOfNextDay - 1
+		perDayTimeRange[i] = interval
+	}
+
+	perDayTimeRange[numOfDays-1].End = end
+
+	return perDayTimeRange
+}
+
+func numOfPlans(start, end model.Time) int {
+	// rounding down start to start of the day
+	if start%model.Time(millisecondPerDay) != 0 {
+		start = model.Time((int64(start) / millisecondPerDay) * millisecondPerDay)
+	}
+
+	// rounding up end to end of the day
+	if end%model.Time(millisecondPerDay) != 0 {
+		end = model.Time((int64(end)/millisecondPerDay)*millisecondPerDay + millisecondPerDay)
+	}
+
+	return int(int64(end-start) / millisecondPerDay)
+}
+
+// groups chunks together by unique label sets i.e all the chunks with same labels would be stored in a group
```

Comment:
Can you comment on why this is done? I don't quite get it.

Reply:
This is just to avoid repeating label names with each chunk. Added the same in comments.